blockstore/apps/bundles/store.py

```diff
@@ -279,28 +334,310 @@ def _create_snapshot(self, bundle_uuid, files):
 
         return snapshot
 
-    def create_snapshot(self, bundle_uuid, paths_to_files):
+
+class DraftRepo:
+    """
+    Similar to SnapshotRepo, except that this class stores mutable Drafts.
+
+    DraftRepos know about SnapshotRepos, but not the other way around.
+    """
+    def __init__(self, snapshot_repo, storage=None):
         """
-        Save the files, create a BundleSnapshot object and save its JSON serialization to storage.
+        A Draft is basically a diff over a Snapshot, so we need a pointer back
+        to where we can get Snapshot data.
         """
-        files = {}
-        for path, data in paths_to_files.items():
-            files[str(path)] = self._save_file(bundle_uuid, path, data)
+        self.snapshot_repo = snapshot_repo
+        self.storage = storage or default_storage
+
+    @classmethod
+    def _data_file_path(cls, draft_uuid, file_path):
+        """Path to a file that we've edited in our Draft."""
+        return f'{draft_uuid}/data/{file_path}'
+
+    @classmethod
+    def _summary_path(cls, draft_uuid):
+        return f'{draft_uuid}/summary.json'
+
+    def _overwrite(self, path, file_obj):
+        # There's gotta be a better way, but for now...
```

Comment:
I mentioned this in reply to a comment on an older version of this, but for since that mention is now hidden, I'll re-post it here where it's still relevant:

> some django-storages backends have an option to allow direct overwrites via `save`, e.g. https://github.com/jschneier/django-storages/search?q=file_overwrite&unscoped_q=file_overwrite

I believe it may also be relatively trivial to subclass django's built-in filestorage implementation to support overwriting for filesystem storage. [django-storages used to include exactly that](https://django-storages.readthedocs.io/en/1.2/backends/overwrite.html) but for some reason no longer does.

Reply:
I'll try the subclassing, but my concern would be that overwrites are in this case necessary for the application to function properly at all. So if we code with the assumption that overwrite will be enabled on any particular backend, then people can shoot themselves in the foot by not properly configuring their storage bucket system. AWS defaults to overwrites = True, but Azure doesn't, etc.

Reply:
Going to punt this since it's likely we'll go with custom backends entirely to enable synchronous commits.